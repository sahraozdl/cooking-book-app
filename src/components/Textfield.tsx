"use client";

import { Field, Input, Label, Textarea } from "@headlessui/react";

function Textfield({
  label,
  id,
  name,
  placeholder,
  defaultValue,
  error,
  type = "text",
  multiline = false,
}: {
  label: string;
  id: string;
  name: string;
  placeholder?: string;
  defaultValue?: string;
  error?: string[];
  type?: string;
  multiline?: boolean;
}) {
  return (
    <Field className="flex flex-col my-4 text-black">
      <Label
        htmlFor={id}
        className="mb-1 text-sm font-semibold text-orange-700"
      >
        {label}
      </Label>
      {multiline ? (
        <Textarea
          id={id}
          name={name}
          placeholder={placeholder}
          aria-describedby={`${id}-error`}
          defaultValue={defaultValue}
          rows={5}
          className={`border rounded px-3 py-2 focus:outline-none focus:ring-2 focus:ring-orange-400 transition bg-white ${
            error ? "border-red-500" : "border-orange-300"
          }`}
        />
      ) : (
        <Input
          id={id}
          name={name}
          type={type}
          placeholder={placeholder}
          aria-describedby={`${id}-error`}
          defaultValue={defaultValue}
          className={`border rounded px-3 py-2 focus:outline-none focus:ring-2 focus:ring-orange-400 transition bg-white ${
            error ? "border-red-500" : "border-orange-300"
          }`}
        />
      )}
      {error && (
        <span id={`${id}-error`} className="text-red-600 text-sm mt-1">
          {error.join(", ")}
        </span>
      )}
    </Field>
  );
}

export default Textfield;
