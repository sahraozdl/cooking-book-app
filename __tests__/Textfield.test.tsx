import { render, screen } from "@testing-library/react";
import Textfield from "@/components/Textfield";

describe("Textfield component", () => {
  it("renders a label and input by default", () => {
    render(<Textfield label="Name" id="name" name="name" />);
    expect(screen.getByLabelText("Name")).toBeInTheDocument();
  });

  it("renders a textarea if multiline is true", () => {
    render(<Textfield label="Description" id="desc" name="desc" multiline />);
    expect(screen.getByRole("textbox")).toBeInTheDocument();
  });

  it("renders error messages if provided", () => {
    render(
      <Textfield
        label="Email"
        id="email"
        name="email"
        error={["Invalid email"]}
      />
    );
    expect(screen.getByText("Invalid email")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Textfield label="Password" id="pwd" name="pwd" />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
