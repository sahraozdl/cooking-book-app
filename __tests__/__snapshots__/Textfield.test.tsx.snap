// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Textfield component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col my-4 text-black"
    data-headlessui-state=""
  >
    <label
      class="mb-1 text-sm font-semibold text-orange-700"
      data-headlessui-state=""
      for="pwd"
      id="headlessui-label-«ra»"
    >
      Password
    </label>
    <input
      aria-labelledby="headlessui-label-«ra»"
      class="border rounded px-3 py-2 focus:outline-none focus:ring-2 focus:ring-orange-400 transition bg-white border-orange-300"
      data-headlessui-state=""
      id="pwd"
      name="pwd"
      type="text"
    />
    <span
      hidden=""
      style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
  </div>
</DocumentFragment>
`;
